File: test/lib/js/bruby_bind/convert.js
(function(){
  'use strict';
  
  let RbValue = BRubyBridge.RbValue;
  let ArgumentError = BRubyBridge.ArgumentError;
  let Convert = BRubyBind.Convert;


  console.log("start-js-bruby_bind-convert");

  
  test("BRubyBind.Convert#convert_call");
  // passed arguments should be converted
  (function(){
    let a;
    let f = function(...args) { a = args; }
    Convert.convert_call(f, undefined, false);
    ensure_result( a[0].equal_to(RbValue.nil) );
    ensure_result( a[1].equal_to(RbValue.false) );
  })();
  // retuned value should be converted
  (function(){
    let f = () => RbValue.nil;
    ensure_result( Convert.convert_call(f) === null );
  })();
  // a thrown rb error should be rethrown as a BRubyBind.RubyError
  (function(){
    let v = RbValue.eval('StandardError.new');
    let f = function() { throw new BRubyBridge.RbError(v); };
    try
    {
      Convert.convert_call(f);
      fail();
    }
    catch (e)
    {
      ensure_result( e instanceof BRubyBind.RubyError );
      ensure_result( e.rb_value === v );
    }
  })();
  // a thrown js error (JavaScript::Error) should be rethrown its js counterpart
  (function(){
    let e = new TypeError;
    global.t = function() { throw e };
    let c = () => RbValue.eval("BRubyBridge::JSValue.global.call('t')");
    try
    {
      Convert.convert_call(c);
      fail();
    }
    catch (ce)
    {
      ensure_result( ce instanceof TypeError );
      ensure_result( ce === e );
    }
    delete global.t;
  })();
  // a thrown basic js error (JavaScript::BasicException) should be rethrown its js counterpart
  (function(){
    global.t = function() { throw 2 };
    let c = () => RbValue.eval("BRubyBridge::JSValue.global.call('t')");
    try
    {
      Convert.convert_call(c);
      fail();
    }
    catch (ce)
    {
      ensure_result( ce === 2 );
    }
    delete global.t;
  })();


  test("BRubyBind.Convert#rb_to_js");
  // passing js object should return the js object
  ensure_result( Convert.rb_to_js(undefined) === undefined );
  // passing rb nil should return js null
  ensure_result( Convert.rb_to_js(RbValue.eval('nil')) === null );
  // passing rb false should reutrn js false
  ensure_result( Convert.rb_to_js(RbValue.eval('false')) === false );
  // passing rb true should reutrn js true
  ensure_result( Convert.rb_to_js(RbValue.eval('true')) === true );
  // passing a rb Numeric should return equal js number
  ensure_result( Convert.rb_to_js(RbValue.eval('1.1')) === 1.1 );
  // passing rb object that responds to #to_int should return equal js number
  rb_eval('$c = Class.new { def to_int; 1; end }')
  ensure_result( Convert.rb_to_js(RbValue.eval('$c.new')) === 1 );
  rb_eval('$c = nil');
  // passing rb String should return an equal js string
  ensure_result( Convert.rb_to_js(RbValue.eval("'abc'")) === 'abc' );
  // passing rb Symbol should return an equivalent js string
  ensure_result( Convert.rb_to_js(RbValue.eval(":abc")) === 'abc' );
  // passing rb object that responds to #to_str should return an equal js string
  rb_eval("$c = Class.new { def to_str; 'abc'; end }");
  global.v = Convert.rb_to_js(RbValue.eval('$c.new'));
  ensure_result( Convert.rb_to_js(RbValue.eval('$c.new')) === 'abc' );
  rb_eval('$c = nil');
  // passing a place holder for js Undefined should return js Undefined
  (function(){
    let v = RbValue.get('JavaScript::Undefined').send('new');
    ensure_result( Convert.rb_to_js(v) === undefined );
  })();
  // passing a rb object that responds to #js_value should return that JSValue
  (function(){
    let o = ({});
    let v = RbValue.get('JavaScript::Object').send('new', o);
    ensure_result( Convert.rb_to_js(v) === o );
  })();
  // passing a rb object that alrady has a RbWrapper built for it should
  //   return the previously built RbWrapper
  (function(){
    let o = RbValue.get('Object').send('new');
    let w1 = Convert.rb_to_js(o);
    ensure_result( w1 instanceof BRubyBind.RubyObject );
    let w2 = Convert.rb_to_js(o);
    ensure_result( w2 === w1 );
  })();
  // passing a rb object that responds to #to_proc
  //   should reutrn a BRubyBind.RubyClosure
  (function(){
    let o = RbValue.eval('Proc.new {}');
    ensure_result( Convert.rb_to_js(o) instanceof BRubyBind.RubyClosure );
  })();
  // passing a rb object that responds to #new
  //   should return a BRubyBind.RubyClass
  (function(){
    let o = RbValue.eval('Class.new');
    ensure_result( Convert.rb_to_js(o) instanceof BRubyBind.RubyClass );
  })();
  // passing a rb object that responds to #exception (but not to #new)
  //   should return an a BRubyBind.RubyError
  (function(){
    let o = RbValue.eval('StandardError.new');
    ensure_result( Convert.rb_to_js(o) instanceof BRubyBind.RubyError );
    o = RbValue.eval('StandardError');
    ensure_result( Convert.rb_to_js(o) instanceof BRubyBind.RubyClass );
  })();
  // passing any other rb object should reutrn a BRubyBind.RubyObject
  (function(){
    let o = RbValue.get('Object').send('new');
    let w = Convert.rb_to_js(o);
    ensure_result( w instanceof BRubyBind.RubyObject );
  })();


  test("BRubyBind.Convert#js_to_rb");
  // passing js undefined should return a rb value of nil
  ensure_result( Convert.js_to_rb(undefined).equal_to(RbValue.nil) )
  // passing js true should return a rb value of true
  ensure_result( Convert.js_to_rb(true).equal_to(RbValue.true) )
  // the reminding possibilities will be tested on the rb side


  console.log("end-js-bruby_bind-convert");
  
})();

File: test/lib/rb/bruby_bind/convert.rb
JSValue =  BRubyBridge::JSValue
Convert = BRubyBind::Convert

puts "start-rb-bruby_bind-convert"


test("BRubyBind::Convert#js_to_rb")
# passing a rb object should return the rb object
ensure_result( Convert.js_to_rb(nil) == nil )
# passing js undefined should return rb nil
ensure_result( Convert.js_to_rb(JSValue.undefined) == nil )
# passing js null should return rb nil
ensure_result( Convert.js_to_rb(JSValue.null) == nil )
# passing js false should return rb false
ensure_result( Convert.js_to_rb(JSValue.false) == false )
# passing js true should reutrn rb true
ensure_result( Convert.js_to_rb(JSValue.true) == true )
# passing a js number rounded to integer should return an equal rb Integer
n = Convert.js_to_rb(JSValue.number(1.0))
ensure_result( n.is_a?(Integer) )
ensure_result( n == 1 )
n = nil
# passing a js number with a non zero decimal value should return an equal rb Float
n = Convert.js_to_rb(JSValue.number(1.5))
ensure_result( n.is_a?(Float) )
ensure_result( n == 1.5 )
n = nil
# passing a js string should return an equal rb String
ensure_result( Convert.js_to_rb(JSValue.string('a')) == 'a' )
# passing a place holder for a rb Symbol should return an equal rb Symbol
v = JSValue["BRubyBind.RubySymbol"].new(JSValue.string('a'))
ensure_result( Convert.js_to_rb(v) == :a )
v = nil
# passing a place holder for a rb Float should return an equal rb Float
v = JSValue["BRubyBind.RubyFloat"].new(JSValue.number(1))
o = Convert.js_to_rb(v)
ensure_result( o.is_a?(Float) )
ensure_result( o == 1.0 )
v, o = nil
# passing a place holder for a rb Integer should return an equal rb Integer
v = JSValue["BRubyBind.RubyInteger"].new(JSValue.number(1.9))
o = Convert.js_to_rb(v)
ensure_result( o.is_a?(Integer) )
ensure_result( o == 1 )
v, o = nil
# passing a js object that has a rb_value property should return that RbValue
v = Object.new
w = JSValue["BRubyBind.RubyObject"].new(v)
r = Convert.js_to_rb(w)
ensure_result( r == v )
v, w, r = nil
# passing a js object that alrady has a JSWrapper built for it should
#   return the previously built JSWrapper
v = JSValue['Object'].new
w1 = Convert.js_to_rb(v)
ensure_result( w1.is_a?(JavaScript::Object) )
w2 = Convert.js_to_rb(v)
ensure_result( w2.equal?(w1) )
v, w1, w2 = nil
# passing js Error should return rb JavaScipt::Error 
w = Convert.js_to_rb(JSValue['Error'])
ensure_result( w.equal?(JavaScript::Error) )
w = nil
# passing a js subclass of Error should return a rb subclass of JavaScipt::Error 
v = JSValue.global['ReferenceError']
JSValue.global.set('tt', v)
w = Convert.js_to_rb(v)
ensure_result( w < JavaScript::Error )
ensure_result( w.js_value == v )
w = nil
# passing a js instance of Error should return an rb instance of JavaScipt::Error
v = JSValue['Error'].new
w = Convert.js_to_rb(v)
ensure_result( w.instance_of?(JavaScript::Error) )
ensure_result( w.js_value == v )
v, w = nil
# passing a js instance of a subclass of Error
#   should return a rb instance a subclass of JavaScipt::Error
v = JSValue['TypeError'].new
w = Convert.js_to_rb(v)
ensure_result_not( w.instance_of?(JavaScript::Error) )
ensure_result( w.is_a?(JavaScript::Error) )
ensure_result( w.js_value == v )
v, w = nil
# passing a js function should reutrn a JavaScript::Function
v = JSValue['Function'].new
w = Convert.js_to_rb(v)
ensure_result( w.is_a?(JavaScript::Function) )
ensure_result( w.js_value == v )
v, w = nil
# passing any other js object should reutrn a JavaScript::Object
v = JSValue['Object'].new
w = Convert.js_to_rb(v)
ensure_result( w.is_a?(JavaScript::Object) )
ensure_result( w.js_value == v )
v, w = nil


test("BRubyBind::Convert#rb_to_js")
# passing rb nil should return a js value of null
ensure_result( Convert.rb_to_js(nil).null? )
# passing rb true should return a js value of true
ensure_result( Convert.rb_to_js(true).true? )
# the reminding possibilities will be tested on the js side


puts("end-rb-bruby_bind-convert")